### DISCUSSION TITLE
Is it possible to control when a FutureProvider completes?

### DISCUSSION THREAD

**Discussion by jibbers42** [rrousselGit/riverpod#1632] [Q&A] [answered]
I have a FutureProvider that I rather not complete if a dependency is certainly not valid yet.
```dart
final sdkFacadeProvider = FutureProvider<SdkFacade>((ref) async {
  final apiKey =
      ref.watch(settingsProvider.select((settings) => settings.apiKey));

  // If apiKey blank then don't complete future..
  if (apiKey == '') {
    ???
  }

  // Have to init sdk
  await Sdk.intializeSdk();

  final sdkApp = SdkApp(options: SdkOptions(apiKey: apiKey));
  return SdkFacade(sdkApp : sdkApp);
});
```

The sdkFacadeProvider is used by other providers that actually make calls to the api and return data. If the api key is blank, then it surely will return 401 Unauthorized. Can I prevent the Future from completing until I'm happy with the input data?

**Comment by TimWhiting** [comment, marked as answer]
With a completer and listening to the settings provider instead of watching it can be done.

**Reply by jibbers42** [reply]
Oooh... listen to it... I'll check that out, thanks!

**Reply by jibbers42** [reply]
For anyone curious, the code I ended up with is basically:
```dart
final sdkFacadeProvider = FutureProvider<SdkFacade>((ref) async {
  final apiKey = ref.watch(settingsProvider.select((settings) => settings.apiKey));
  final apiKeyCompleter = Completer<SdkFacade>();
  if (apiKey != '') {
    // Have to init sdk
    await Sdk.reset(); // no op if not initialized
    await Sdk.intializeSdk();

    final sdkApp = SdkApp(options: SdkOptions(apiKey: apiKey));
    apiKeyCompleter.complete(SdkFacade(sdkApp: sdkApp));
  }
  return apiKeyCompleter.future;
});
```

This seems to work. I kept the `watch` to produce a new future out of sdkFacadeProvider when the key changes (and isn't blank).

I guess futures for blank keys are never completed... I'm not sure if I need to cancel or dispose of those futures in some way or not???

**Reply by TimWhiting** [reply]
> I guess futures for blank keys are never completed... I'm not sure if I need to cancel or dispose of those futures in some way or not???

If you are worried about this you could make the provider return a nullable SdkFacade or choose to complete the future with completeError depending on what you want the UI / providers watching this provider to do.

**Reply by jibbers42** [reply]
> If you are worried about this you could make the provider return a nullable SdkFacade or choose to complete the future with completeError depending on what you want the UI / providers watching this provider to do.

I'm not really worried about it since it's a relatively small app and I can't imagine why this would change more than a handful of times per app run. But I was curious, so that approach to handling it is great to know.
